flight/PiOS/Common/pios_bmi160.c

```diff
@@ -495,165 +495,198 @@ bool PIOS_BMI160_IRQHandler(void)
+	float gyro_z = (int16_t)(bmi160_rec_buf[IDX_GYRO_ZOUT_H] << 8 | bmi160_rec_buf[IDX_GYRO_ZOUT_L]);
+
+	// Apply sensor scaling
+	accel_x *= dev->accel_scale;
+	accel_y *= dev->accel_scale;
+	accel_z *= dev->accel_scale;
+
+	gyro_x *= dev->gyro_scale;
+	gyro_y *= dev->gyro_scale;
+	gyro_z *= dev->gyro_scale;
+
+	/*
+	 * Convert from sensor frame (x: forward y: left z: up) to
+	 * TL convention (x: forward y: right z: down).
+	 * See flight/Doc/imu_orientation.md for more detail
+	 */
+	switch (dev->cfg->orientation) {
+	case PIOS_BMI160_TOP_0DEG:
+		accel_data->x = accel_x;
+		accel_data->y = -accel_y;
+		accel_data->z = -accel_z;
+		gyro_data->x  = gyro_x;
+		gyro_data->y  = -gyro_y;
+		gyro_data->z  = -gyro_z;
+		break;
+	case PIOS_BMI160_TOP_90DEG:
+		accel_data->x = accel_y;
+		accel_data->y = accel_x;
+		accel_data->z = -accel_z;
+		gyro_data->x  = gyro_y;
+		gyro_data->y  = gyro_x;
+		gyro_data->z  = -gyro_z;
+		break;
+	case PIOS_BMI160_TOP_180DEG:
+		accel_data->x = -accel_x;
+		accel_data->y = accel_y;
+		accel_data->z = -accel_z;
+		gyro_data->x  = -gyro_x;
+		gyro_data->y  = gyro_y;
+		gyro_data->z  = -gyro_z;
+		break;
+	case PIOS_BMI160_TOP_270DEG:
+		accel_data->x = -accel_y;
+		accel_data->y = -accel_x;
+		accel_data->z = -accel_z;
+		gyro_data->x  = -gyro_y;
+		gyro_data->y  = -gyro_x;
+		gyro_data->z  = -gyro_z;
+		break;
+	case PIOS_BMI160_BOTTOM_0DEG:
+		accel_data->x = accel_x;
+		accel_data->y = accel_y;
+		accel_data->z = accel_z;
+		gyro_data->x  = gyro_x;
+		gyro_data->y  = gyro_y;
+		gyro_data->z  = gyro_z;
+		break;
+	case PIOS_BMI160_BOTTOM_90DEG:
+		accel_data->x = -accel_y;
+		accel_data->y = accel_x;
+		accel_data->z = accel_z;
+		gyro_data->x  = -gyro_y;
+		gyro_data->y  = gyro_x;
+		gyro_data->z  = gyro_z;
+		break;
+	case PIOS_BMI160_BOTTOM_180DEG:
+		accel_data->x = -accel_x;
+		accel_data->y = -accel_y;
+		accel_data->z = accel_z;
+		gyro_data->x  = -gyro_x;
+		gyro_data->y  = -gyro_y;
+		gyro_data->z  = gyro_z;
+		break;
+	case PIOS_BMI160_BOTTOM_270DEG:
+		accel_data->x = accel_y;
+		accel_data->y = -accel_x;
+		accel_data->z = accel_z;
+		gyro_data->x  = gyro_y;
+		gyro_data->y  = -gyro_x;
+		gyro_data->z  = gyro_z;
+		break;
+	}
 
+	dev->sensor_ready |= SENSOR_ACCEL;
+
+	// Get the temperature
+	// NOTE: We do this down here so the chip-select has some time to go low. Strange things happen
+	// When this is done right after readin the accels / gyros
+	if (dev->temp_interleave_cnt % dev->cfg->temperature_interleaving == 0){
 		if (PIOS_BMI160_ClaimBus() != 0)
-			continue;
+			return 0;
 
-		if (PIOS_SPI_TransferBlock(dev->spi_id, bmi160_tx_buf, bmi160_rec_buf, BUFFER_SIZE) < 0) {
+		uint8_t bmi160_tx_buf[BUFFER_SIZE] = {BMI160_REG_TEMPERATURE_0 | 0x80, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0};
+		if (PIOS_SPI_TransferBlock(dev->spi_id, bmi160_tx_buf, bmi160_rec_buf, 3) < 0) {
 			PIOS_BMI160_ReleaseBus();
-			continue;
+			return 0;
```

Comment:
As above.

Reply:
I've interpreted the comment in front of the if clause, that this has the chance of being occasionally problematic, so we probably might not care that it fails here and there.

Reply:
Actually re-reading the diff, yeah, you're right. It's functionally different that way.